subsys/debug/ppi_trace/ppi_trace.c

```diff
@@ -0,0 +1,213 @@
+#define IS_PAIR(handle) ((u32_t)handle & PAIR_FLAG)
+
+#define GET_CH(handle) ((u32_t)handle & 0xFF)
+#define GET_START_CH(handle) GET_CH(handle)
+#define GET_STOP_CH(handle) (((u32_t)handle >> 8) & 0xFF)
+
+/** @brief Allocate (D)PPI channel. */
+static nrfx_err_t ppi_alloc(u8_t *ch, u32_t evt)
+{
+	nrfx_err_t err;
+#ifdef DPPI_PRESENT
+	if (*PUBLISH_ADDR(evt) != 0) {
+		/* Use mask of one of subscribe registers in the system,
+		 * assuming that all subscribe registers has the same mask for
+		 * channel id.
+		 */
+		*ch = *PUBLISH_ADDR(evt) & DPPIC_SUBSCRIBE_CHG_EN_CHIDX_Msk;
+		err = NRFX_SUCCESS;
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+/** @brief Set task and event on (D)PPI channel. */
+static void ppi_assign(u8_t ch, u32_t evt, u32_t task)
+{
+#ifdef DPPI_PRESENT
+	/* Use mask of one of subscribe registers in the system, assuming that
+	 * all subscribe registers has the same mask for enabling channel.
+	 */
+	*SUBSCRIBE_ADDR(task) = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | (u32_t)ch;
+	*PUBLISH_ADDR(evt) = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | (u32_t)ch;
+#else
+	(void)nrfx_ppi_channel_assign(ch, evt, task);
+#endif
+}
+
+/** @brief Enable (D)PPI channels. */
+static void ppi_enable(u32_t channel_mask)
+{
+#ifdef DPPI_PRESENT
+	(void)nrf_dppi_channels_enable(channel_mask);
+#else
+	(void)nrf_ppi_channels_enable(channel_mask);
+#endif
+}
+
+/** @brief Disable (D)PPI channels. */
+static void ppi_disable(u32_t channel_mask)
+{
+#ifdef DPPI_PRESENT
+	(void)nrf_dppi_channels_disable(channel_mask);
+#else
+	(void)nrf_ppi_channels_disable(channel_mask);
+#endif
+}
+
+/** @brief Allocate GPIOTE channel.
+ *
+ * @param pin Pin.
+ *
+ * @return Allocated channel or -1 if failed to allocate.
+ */
+static int gpiote_channel_alloc(u32_t pin)
+{
+	for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel) {
+		if (!nrf_gpiote_te_is_enabled(channel)) {
+			nrf_gpiote_task_configure(channel, pin,
+						  NRF_GPIOTE_POLARITY_TOGGLE,
+						  NRF_GPIOTE_INITIAL_VALUE_LOW);
+			nrf_gpiote_task_enable(channel);
+			return channel;
+		}
+	}
+
+	return -1;
+}
+
+void *ppi_trace_config(u32_t pin, u32_t evt)
+{
+	int err;
+	u32_t task;
+	int gpiote_ch;
+	nrf_gpiote_tasks_t task_id = NRF_GPIOTE_TASKS_OUT_0;
+	u8_t ppi_ch;
+
+	err = ppi_alloc(&ppi_ch, evt);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to allocate PPI channel.");
+		return NULL;
+	}
+
+	gpiote_ch = gpiote_channel_alloc(pin);
+	if (gpiote_ch < 0) {
+		LOG_ERR("Failed to allocate GPIOTE channel.");
+		return NULL;
+	}
+
+	task_id += gpiote_ch*sizeof(u32_t);
+	task = nrf_gpiote_task_addr_get(task_id);
+	ppi_assign(ppi_ch, evt, task);
+
+	return HANDLE_ENCODE(ppi_ch);
+}
+
+void *ppi_trace_pair_config(u32_t pin, u32_t start_evt, u32_t stop_evt)
+{
+	int err;
+	u32_t task;
+	u8_t start_ch;
+	u8_t stop_ch;
+	int gpiote_ch;
+
+	if (!IS_ENABLED(GPIOTE_FEATURE_SET_PRESENT)) {
```

Comment:
Will it actually work? `GPIOTE_FEATURE_SET_PRESENT` is defined without a value.

Why not use just:
```
#if !defined(GPIOTE_FEATURE_SET_PRESENT) || \
    !defined(GPIOTE_FEATURE_CLR_PRESENT)
	__ASSERT(0, "Function not supported on this platform.");
	return NULL;
#else
	...
```
?

Reply:
fixed.